Fixes to merged 3.11.0 driver

avelanarius opened the pull request:
This PR consists of necessary changes to the 3.11.0 branch after merging from upstream.

**First commit** fixes protocol version downgrading. Previously, if an `OPTIONS` packet was sent with a wrong protocol version,
`onOptionsResponse` handler method would return a generic `TransportException` exception. This commit implements a similar approach to `onStartupResponse()` by throwing `UnsupportedProtocolVersion` when the exception is of that kind.

This fixes a problem with protocol version downgrading - before this change, the driver could not correctly downgrade the protocol version. By returning a properly typed Exception, the driver can correctly retry with a lower protocol version. The reason protocol version downgrading was broken in our fork and not in upstream is that we send `OPTIONS` packet before `STARTUP`, while upstream does it in reverse.

**Second commit** introduces `ProtocolVersion.DEFAULT` - a default protocol version to be used while connecting to a cluster. Previously `NEWEST_SUPPORTED` was used as a default, but 3.11.0 introduced support for protocol version V5, which is not supported in Scylla. Such default would cause the driver to always downgrade the version upon connecting. Therefore the new default (`ProtocolVersion.DEFAULT`) is set to `V4`.

avelanarius committed:
Throw UnsupportedProtocolVersion handling OPTIONS

Previously, if an OPTIONS packet was sent with a wrong protocol version,
onOptionsResponse handler method would return a generic
TransportException exception. This patch implements a similar approach
to onStartupResponse() by throwing UnsupportedProtocolVersion when
the exception is of that kind.

This fixes a problem with protocol version downgrading - before this
change, the driver could not correctly downgrade the protocol version
and failed with:

```
com.datastax.driver.core.exceptions.NoHostAvailableException:
All host(s) tried for query failed (tried: /127.0.1.1:38223
(com.datastax.driver.core.exceptions.TransportException:
[/127.0.1.1:38223] Got ERROR response message from server to an
OPTIONS message: Beta version of the protocol used (5/v5-beta),
but USE_BETA flag is unset))
```

By returning a properly typed Exception, the driver can correctly retry
with a lower protocol version. The reason protocol version downgrading
was broken in our fork and not in upstream is that we send OPTIONS
packet before STARTUP, while upstream does it in reverse.

avelanarius committed:
Introduce ProtocolVersion.DEFAULT

Before this change, ProtocolVersion.NEWEST_SUPPORTED was used as a
default version while connecting to a cluster. DataStax Driver 3.11.0
introduced support for protocol version V5 and accordingly set
NEWEST_SUPPORTED to v5. This is a problem for Scylla, as it only
supports v4 protocol version causing the driver to always downgrade
the protocol version upon connecting. By introducing DEFAULT (set to v4)
the driver now connects with older protocol version, while maintaining
support for higher protocol versions (NEWEST_SUPPORTED set to v5).

haaawk commented:
LGTM in general. Two questions though:

1. Is it ok to still use `NEWEST_SUPPORTED` in `driver-core/src/main/java/com/datastax/driver/core/DataTypeClassNameParser.java` `getNameAndTypeParameters`?
2. I can see that `NEWEST_SUPPORTED` is used in some tests. Would it be better to test DEFAULT there?
```
driver-core/src/test/java/com/datastax/driver/core/ClusterInitTest.java:                  ProtocolVersion.NEWEST_SUPPORTED.minCassandraVersion().toString())
driver-core/src/test/java/com/datastax/driver/core/DataTypeClassNameParserTest.java:  private ProtocolVersion protocolVersion = ProtocolVersion.NEWEST_SUPPORTED;
driver-core/src/test/java/com/datastax/driver/core/DataTypeTest.java:  ProtocolVersion protocolVersion = ProtocolVersion.NEWEST_SUPPORTED;
driver-core/src/test/java/com/datastax/driver/core/NettyOptionsTest.java:    ProtocolVersion version = ProtocolVersion.NEWEST_SUPPORTED;
driver-core/src/test/java/com/datastax/driver/core/OPPTokenFactoryTest.java:    return factory.deserialize(buffer, ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/PreparedStatementTest.java:    assertThat(bs.getRoutingKey(ProtocolVersion.NEWEST_SUPPORTED, CodecRegistry.DEFAULT_INSTANCE))
driver-core/src/test/java/com/datastax/driver/core/PreparedStatementTest.java:    assertThat(bs.getRoutingKey(ProtocolVersion.NEWEST_SUPPORTED, CodecRegistry.DEFAULT_INSTANCE))
driver-core/src/test/java/com/datastax/driver/core/ProtocolBetaVersionTest.java:    // Note: when the driver's ProtocolVersion.NEWEST_SUPPORTED will be incremented to V6 or higher
driver-core/src/test/java/com/datastax/driver/core/ProtocolBetaVersionTest.java:    // Note: when the driver's ProtocolVersion.NEWEST_SUPPORTED will be incremented to V6 or higher
driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java:    if (protocolVersion.compareTo(ProtocolVersion.NEWEST_SUPPORTED) >= 0) {
driver-core/src/test/java/com/datastax/driver/core/StatementIdempotenceTest.java:    when(protocolOptions.getProtocolVersion()).thenReturn(ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/TypeCodecAssert.java:  private ProtocolVersion version = ProtocolVersion.NEWEST_SUPPORTED;
driver-core/src/test/java/com/datastax/driver/core/TypeCodecTest.java:            ProtocolVersion.NEWEST_SUPPORTED,
driver-core/src/test/java/com/datastax/driver/core/TypeCodecTest.java:            .deserialize(ByteBuffer.allocate(0), ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/TypeCodecTest.java:            ProtocolVersion.NEWEST_SUPPORTED,
driver-core/src/test/java/com/datastax/driver/core/TypeCodecTest.java:            .deserialize(ByteBuffer.allocate(0), ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:    when(protocolOptions.getProtocolVersion()).thenReturn(ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:              .serialize("This is some sample text", ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:                  ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:                  ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:                  ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:      ByteBuffer routingKey = TypeCodec.bigint().serialize(33L, ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:          TypeCodec.bigint().serialize(42L, ProtocolVersion.NEWEST_SUPPORTED);
driver-core/src/test/java/com/datastax/driver/core/policies/TokenAwarePolicyTest.java:          TypeCodec.varchar().serialize("hello_world", ProtocolVersion.NEWEST_SUPPORTED);
```

avelanarius committed:
Use non-hardcoded protocol version in Parser

Instead of always using ProtocolVersion.NEWEST_SUPPORTED in
DataTypeClassNameParser, use the actual protocolVersion.

avelanarius committed:
Add protocol version down-negotiation test

Add a new integration test which creates a connection to a Cluster
with NEWEST_SUPPORTED protocol version (not the default DEFAULT) to 
check if the driver can successfully down-negotiate the protocol version.

avelanarius commented:
v2:
- In `DataTypeClassNameParser` instead of hardcoding `ProtocolVersion.NEWEST_SUPPORTED` now it's using the proper selected protocol version.
- Changed `ProtocolVersion.NEWEST_SUPPORTED` to `ProtocolVersion.DEFAULT` in all files where it made sense.
- Added a new integration test which creates a connection with a high protocol version (`NEWEST_SUPPORTED`) and checks if the driver can downgrade it and still successfully connect. This required adding a new test-only global variable `shouldAlwaysNegotiateProtocolVersion`, because if a user explicitly sets a protocol version, the driver will not try downgrading the connection (the purpose of this new test). Before this integration test, such code path was never tested.

haaawk reviewed (approved):
LGTM
